### PR TITLE
[FIX] Pad feature vectors to the same size in xgboost cost model

### PR DESCRIPTION
**Pull request by tkonolige**
I'm hitting this issue when tuning `dense_pack.x86`.

@tqchen @merrymercy

### CODE REVIEW

**Review by merrymercy — COMMENTED** [review]
I do not think this should be handled here. We should fix it in feature extraction and make sure the feature extraction always outputs vectors with the same length.

**Comment by tkonolige** [comment]
@merrymercy as far as I can tell, feature extraction does not guarantee that feature vectors are the same length. Looking at `src/autotvm/touch_extractor.cc` https://github.com/apache/tvm/blob/main/src/autotvm/touch_extractor.cc#L335 the length of the the feature vector depends on the number of itervars in the program. If the schedule used applies a different number of splits based on some autotvm config, then the number of itervars will not be consistent.

It seems that we have to add padding somewhere. We can either add it in the xgb tuner or in the feature extractor. If we add it to the feature extractor, then we need to guess the maximum number of itervars allowed which could either lead to lots of wasted space in the feature vector or failures when there are more itervars than expected. To me it seems easier to apply the fix in the tuner.

**Comment by merrymercy** [comment]
You are right. Sorry, I thought this was auto_scheduler instead of autotvm.

**Review by AndrewZhaoLuo — COMMENTED** [review]
LGTM. How hard is it to add a test for this?

File: tests/python/unittest/test_autotvm_xgboost_model.py [hunk 1]
```diff
@@ -43,6 +43,8 @@ def test_fit():
 
     upper_model.fit(xs, ys, plan_size=32)
 
+    upper_model.predict(np.ones(12))
```
Comment (tkonolige) [inline tests/python/unittest/test_autotvm_xgboost_model.py:46]: Surprisingly, xgboost seems to be fine doing predictions on vectors longer than what it was trained on.

File: python/tvm/testing/autotvm.py [hunk 2]
```diff
@@ -62,9 +62,12 @@ def matmul(N, L, M, dtype):
 
     # schedule according to config
     yo, yi = cfg["tile_y"].apply(s, C, y)
-    xo, xi = cfg["tile_x"].apply(s, C, x)
+    if cfg["tile_x"].size[-1] > 1:
+        xo, xi = cfg["tile_x"].apply(s, C, x)
```
Comment (AndrewZhaoLuo) [inline python/tvm/testing/autotvm.py:66]: Can you give some context on this in the code? Is this to hit the previously failing case?

File: python/tvm/autotvm/tuner/xgboost_cost_model.py [hunk 3]
```diff
@@ -243,18 +243,27 @@ def fit_log(self, records, plan_size, min_seed_records=500):
...
+                continue
+            x, _ = res.value
+            fea_len = max(fea_len, x.shape[0])
 
         xs, ys = [], []
         for res in result:
             if res.status != StatusKind.COMPLETE:
                 continue
             x, y = res.value
-            if len(x) == fea_len:
```
Comment (AndrewZhaoLuo) [inline python/tvm/autotvm/tuner/xgboost_cost_model.py:255]: Hmm, I think previously glossed over the significance of this. Before, it would only use features from those with the same shape as the first entry to fit the model so basically loops with the same number of axis. So this is actually quite a major change is it not? 

The model is at the task level, so how did it handle configs with different size features between training rounds?

File: python/tvm/testing/autotvm.py (hunk 2 above)
Comment (tkonolige) [reply python/tvm/testing/autotvm.py:66]: Yeah, this is to hit the failing test case. It will conditionally split one of the loops so that some configurations have n itervars and some have n+1.

Comment (AndrewZhaoLuo) [reply python/tvm/testing/autotvm.py:66]: Can you add a comment

**Review by AndrewZhaoLuo — APPROVED** [review]

File: python/tvm/testing/autotvm.py (hunk 2 above)
Comment (tkonolige) [reply python/tvm/testing/autotvm.py:69]: done
